adr/context.py

```diff
@@ -0,0 +1,287 @@
+    [['--sort-key'],
+     {'type': int,
+      'dest': 'sort_key',
+      'default': 4,
+      'help': "Key to sort on (int, 0-based index)",
+      }]
+]
+"""
+These are commonly used arguments which can be re-used. They are shared to
+provide a consistent CLI across recipes/queries.
+"""
+
+
+class ZeroContextEvaluator(ExpressionEvaluator):
+    """
+    Zero context evaluator is used in conjunction with extract_context_names
+    to extract argument context as a token value
+    """
+    @staticmethod
+    def update_missing(dest_dict, source_dict):
+        """
+        Update missing key from source dict to destination dict
+        :param dict dest_dict: destination dist to be updated
+        :param dict source_dict: source dict used to update destination
+        :return: None
+        """
+        for k, v in source_dict.items():
+            if k not in dest_dict:
+                dest_dict[k] = v
+
+    def __init__(self):
+        # Init evaluator with fake empty context
+        super(ZeroContextEvaluator, self).__init__({})
+        # update infix_rules and prefix_rules with override function
+        ZeroContextEvaluator.update_missing(self.infix_rules,
+                                            super(ZeroContextEvaluator, self).infix_rules)
+        ZeroContextEvaluator.update_missing(self.prefix_rules,
+                                            super(ZeroContextEvaluator, self).prefix_rules)
+        # Init empty context set
+        self.contexts = set()
+
+    @prefix("identifier")
+    def identifier(self, token, pc):
+        """
+        Get corresponding value for token
+        In this function, we return the same token value without using context
+        :param token: token object
+        :param pc: parse context
+        :return: same token value
+        """
+        self.contexts.add(token.value)
+        return token.value
+
+    def parse(self, expression):
+        """
+        Parse current expression, return contexts inside this expression
+        :param str expression: expression
+        :return: set of contexts
+        """
+        self.contexts.clear()
+        super(ZeroContextEvaluator, self).parse(expression)
+        return self.contexts
+
+
+G_ZEROCONTEXT_EVALUATOR = ZeroContextEvaluator()
+
+
+def extract_context_names(query):
+    """
+    Extract contexts list from query object
+    This is to mimic json-e render() function
+    Given using only $eval in query
+    :param dict query: query object to extract context
+    :return: set of context names
+    """
+    contexts = set()
+    if isinstance(query, dict):
+        for k, v in query.items():
+            if k == "$eval":
+                contexts.update(G_ZEROCONTEXT_EVALUATOR.parse(v))
+            else:
+                contexts.update(extract_context_names(v))
+    elif isinstance(query, list):
+        for v in query:
+            contexts.update(extract_context_names(v))
+
+    return contexts
+
+
+def _get_destination(definition):
```

Comment:
Let's just take the guesswork out of this and turn COMMON_CONTEXTS into a dict:

    COMMON_CONTEXT = {
        'name': { ... definition ... },
        ...
    }

I think doing this would make this and `_get_context_definition` unnecessary. When we load the common context, we can automatically set its "dest" value to the name.

Reply:
Yes, I updated the `COMMON_CONTEXTS` and other definitions, so I can eliminate many funtions :-) Please refer to #85